src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,46 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
 
+    public void push(T value) {
+        top = new Node<>(value, top);
+        size++;
     }
 
     public T peek() {
-        return null;
+        checkSize();
+        return top.value;
     }
 
     public T pop() {
-        return null;
+        checkSize();
+        T value = top.value;
+        top = top.top;
+        size--;
+        return value;
+    }
+
+    private void checkSize() {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> top;
```

Comment:
```suggestion
        private Node<T> previous;
```
`top` is the top node in the stack, but each node has the link to the previous one, right?